File: Python_Scripts/py39/Photosynthetic/Python_Version/choose_monomer.py
class K:
    def __init__(self):
        self.current_dirs = None
        self.current_dirs = set(dir(self))
        pass
    
    def __str__(self):
        string = ''
        for each in set(dir(self))-self.current_dirs:
            param = f'self.{each}'
            string += f'{each}: {eval(param)}\n'
        return string

class Monomer:
    tags = [i+1 for i in range( 7 )] +[-1]
    def __repr__(self):
        return f'{self.name}: {self.moles:.2e} Mol'
    def __str__(self):
        return self.name
    def __init__(self, name, conc,conc_type, kdummy = 1e4):
        self.K = K()
        if name in ('test', '-1'):
            self.name   = 'test'           # store the monomer for graphs etc
            self.tag    = '-1'             # numerical tag to make picking the Rij easier
            self.vmax   = 2                # max number of vinyl groups in monomer
            self.pmax   = 2                # max number of radicals on a monomer unit
            self.nmax   = 2 * self.vmax    # max number of bonds per monomer unit
            self.M      = 50               # g/mol
            self.dens   = 1                # g/mL at 25 °C
            self.K.init = 7*kdummy
            self.K.prop = kdummy           # the propogation rate of this monomer
            self.K.tr   = kdummy           # the recombination termination rate of this monomer
            self.K.td   = kdummy           # the disproportionation termination rate of this monomer
        elif name in ('BPAEDA', '1'):
            self.name   = 'BPAEDA'
            self.tag    = '1'
            self.vmax   = 2
            self.pmax   = 2
            self.nmax   = 2 * self.vmax
            self.M      = 688
            self.dens   = 1.15
            self.K.init = 7*kdummy
            self.K.prop = 7*kdummy
            self.K.tr   = 100*kdummy
            self.K.td   = 100*kdummy
        elif name in ('TEGDMA', '2'):
            self.name   = 'TEGDMA'
            self.tag    = '2'
            self.vmax   = 2
            self.pmax   = 2
            self.nmax   = 2 * self.vmax
            self.M      = 286.32
            self.dens   = 1.092
            self.K.init = 7*kdummy
            self.K.prop = 7*kdummy
            self.K.tr   = 100*kdummy
            self.K.td   = 100*kdummy
        elif name in ('bisGMA', '3'):
            self.name   = 'bisGMA'
            self.tag    = '3'
            self.vmax   = 2
            self.pmax   = 2
            self.nmax   = 2 * self.vmax
            self.M      = 512.59
            self.dens   = 1.161
            self.K.init = 5*kdummy
            self.K.prop = 1*kdummy
            self.K.tr   = 100*kdummy
            self.K.td   = 100*kdummy
        elif name in ('TEGDVE','4'):
            self.name   = 'TEGDVE'
            self.tag    = '4'
            self.vmax   = 2
            self.pmax   = 2
            self.nmax   = 2 * self.vmax
            self.M      = 202.25
            self.dens   = 0.99
            self.K.init = 5*kdummy
            self.K.prop = 1*kdummy
            self.K.tr   = 100*kdummy
            self.K.td   = 100*kdummy
        elif name in ('NPM', '5'):
            self.name   = 'NPM' # N-Propylmaleimide
            self.tag    = '5'
            self.vmax   = 1
            self.pmax   = 1
            self.nmax   = 2 * self.vmax
            self.M      = 139.15
            self.dens   = 1.112
            self.K.init = 3*kdummy
            self.K.prop = 3*kdummy
            self.K.tr   = 100*kdummy
            self.K.td   = 100*kdummy
        elif name in ('PEGDA', '6'):
            self.name   = 'PEGDA'
            self.header = [self.name, f'{self.name}_radical']
            self.tag    = '6'
            self.vmax   = 2
            self.pmax   = 2
            self.nmax   = 2 * self.vmax
            self.M      = 360.18
            self.dens   = 1.11
            self.K.init = 5.787e+05 #  L/mol s %% rmg
            self.K.prop = 2.878e+04 # + 227.6 + 250.5 % L/mol s %% rmg 1!: R 173, 174, 196
            self.K.td   = 4.3955e7
            self.K.tr   = 4.3955e7
            # monomer.K.tr   = 8.658e+8  % L/mol s %% rmg
            # monomer.K.td   = 1.21e+9   % L/mol s %% rmg RMG 1: R 291
        elif name in ('HDDA', '7'):
            self.name   = 'HDDA'
            self.tag    = '7'
            self.vmax   = 2
            self.pmax   = 2
            self.nmax   = 2 * self.vmax
            self.M      = 226.27
            self.dens   = 1.01
            self.K.init = 6.1197e4
            self.K.prop = 6.1197e4
            self.K.tr   = 4.3955e7
            self.K.td   = 4.3955e7
        else:
            raise ValueError(f'This monomer ({name}) has not yet been added, did you type it right dummy?')
        self.K.Rij  = {f'R{self.tag}{i}' : 1 for i in Monomer.tags}


        # change to function on species class
        if conc_type == 'mol':
            self.moles = conc
        elif conc_type == 'g':
            self.moles = conc/self.M
        else:
            raise ValueError(f"That input type {conc_type} is not supported")


        self.headers = (self.name, f'{self.name}-radical')






if __name__ == "__main__":
    import run_main

File: Python Scripts/py39/Photosynthetic/Reference/choose_monomer.m
function [monomer] = choose_monomer(input,conc,type)
%CHOOSE_MONOMER returns a struct containing the specifications for the
%monomer by inputting the name of the wanted monomer
%   Molarmass and densitys from Sigma Aldrich
%   RMG sources from estimation software https://rmg.mit.edu
%   RMG 1: reduced-PEGDA-230-prop | indege/ !outedge

kdummy = 1e4;

if strcmp(input, 'test') || strcmp(input, '-1')
    %%
    monomer.name   = 'test';           % store the monomer for graphs etc
    monomer.tag    = '-1';             % numerical tag to make picking the Rij easier
    monomer.vmax   = 2;                % max number of vinyl groups in monomer
    monomer.pmax   = 2;                % max number of radicals on a monomer unit
    monomer.nmax   = 2 * monomer.vmax; % max number of bonds per monomer unit
    monomer.M      = 50;               % g/mol
    monomer.dens   = 1;                % g/mL at 25 °C
    monomer.K.Rii  = 1;                % always 1 as reactivity with itself is 1
    monomer.K.Rij  = 1;                % the reactivity rate of the reaction between radical i and monomer j
    monomer.K.init = 7*kdummy;
    monomer.K.prop = kdummy;           % the propogation rate of this monomer
    monomer.K.tr   = kdummy;           % the recombination termination rate of this monomer
    monomer.K.td   = kdummy;           % the disproportionation termination rate of this monomer
elseif strcmp(input, 'BPAEDA') || strcmp(input, '1')
    %%
    monomer.name   = 'BPAEDA';
    monomer.tag    = '1';
    monomer.vmax   = 2;
    monomer.pmax   = 2;
    monomer.nmax   = 2 * monomer.vmax;
    monomer.M      = 688;
    monomer.dens   = 1.15;
    monomer.K.R11  = 1;
    monomer.K.R12  = 1;
    monomer.K.R13  = 1;
    monomer.K.R14  = 1;
    monomer.K.R15  = 1;
    monomer.K.R16  = 1;
    monomer.K.R17  = 1;
    monomer.K.init = 7*kdummy;
    monomer.K.prop = 7*kdummy;
    monomer.K.tr   = 100*kdummy;
    monomer.K.td   = 100*kdummy;
elseif strcmp(input, 'TEGDMA')  || strcmp(input, '2')
    %%
    monomer.name   = 'TEGDMA';
    monomer.tag    = '2';
    monomer.vmax   = 2;
    monomer.pmax   = 2;
    monomer.nmax   = 2 * monomer.vmax;
    monomer.M      = 286.32;
    monomer.dens   = 1.092;
    monomer.K.R22  = 1;
    monomer.K.R21  = 1;
    monomer.K.R23  = 1;
    monomer.K.R24  = 1;
    monomer.K.R25  = 1;
    monomer.K.R26  = 1;
    monomer.K.R27  = 1;
    monomer.K.init = 7*kdummy;
    monomer.K.prop = 7*kdummy;
    monomer.K.tr   = 100*kdummy;
    monomer.K.td   = 100*kdummy;
elseif strcmp(input, 'bisGMA') || strcmp(input, '3')
    %%
    monomer.name   = 'bisGMA';
    monomer.tag    = '3';
    monomer.vmax   = 2;
    monomer.pmax   = 2;
    monomer.nmax   = 2 * monomer.vmax;
    monomer.M      = 512.59;
    monomer.dens   = 1.161;
    monomer.K.R33  = 1;
    monomer.K.R31  = 1;
    monomer.K.R32  = 1;
    monomer.K.R34  = 1;
    monomer.K.R35  = 1;
    monomer.K.R36  = 1;
    monomer.K.R37  = 1;
    monomer.K.init = 5*kdummy;
    monomer.K.prop = 1*kdummy;
    monomer.K.tr   = 100*kdummy;
    monomer.K.td   = 100*kdummy;
elseif strcmp(input, 'TEGDVE') || strcmp(input, '4')
    %%
    monomer.name   = 'TEGDVE';
    monomer.tag    = '4';
    monomer.vmax   = 2;
    monomer.pmax   = 2;
    monomer.nmax   = 2 * monomer.vmax;
    monomer.M      = 202.25;
    monomer.dens   = 0.99;
    monomer.K.R44  = 1;
    monomer.K.R41  = 1;
    monomer.K.R42  = 1;
    monomer.K.R43  = 1;
    monomer.K.R45  = 1;
    monomer.K.R46  = 1;
    monomer.K.R47  = 1;
    monomer.K.init = 5*kdummy;
    monomer.K.prop = 1*kdummy;
    monomer.K.tr   = 100*kdummy;
    monomer.K.td   = 100*kdummy;
elseif strcmp(input, 'NPM')    || strcmp(input, '5')
    %%
    monomer.name   = 'NPM'; % N-Propylmaleimide
    monomer.tag    = '5';
    monomer.vmax   = 1;
    monomer.pmax   = 1;
    monomer.nmax   = 2 * monomer.vmax;
    monomer.M      = 139.15;
    monomer.dens   = 1.112;
    monomer.K.R55  = 1;
    monomer.K.R51  = 1;
    monomer.K.R52  = 1;
    monomer.K.R53  = 1;
    monomer.K.R54  = 1;
    monomer.K.R56  = 1;
    monomer.K.R57  = 1;
    monomer.K.init = 3*kdummy;
    monomer.K.prop = 3*kdummy;
    monomer.K.tr   = 100*kdummy;
    monomer.K.td   = 100*kdummy;
elseif strcmp(input, 'PEGDA')  || strcmp(input, '6')
    %%
    monomer.name   = 'PEGDA';
    monomer.tag    = '6';
    monomer.vmax   = 2;
    monomer.pmax   = 2;
    monomer.nmax   = 2 * monomer.vmax;
    monomer.M      = 360.18;
    monomer.dens   = 1.11;
    monomer.K.R66  = 1;
    monomer.K.R61  = 1;
    monomer.K.R62  = 1;
    monomer.K.R63  = 1;
    monomer.K.R64  = 1;
    monomer.K.R65  = 1;
    monomer.K.R67  = 1;
    monomer.K.init = 5.787e+05;  % L/mol s %% rmg
    monomer.K.prop = 2.878e+04; % + 227.6 + 250.5; % L/mol s %% rmg 1!: R 173, 174, 196
    % monomer.K.tr   = 8.658e+8;  % L/mol s %% rmg
    % monomer.K.td   = 1.21e+9;   % L/mol s %% rmg RMG 1: R 291
    monomer.K.tr   = 4.3955e7;
    monomer.K.td   = 4.3955e7;
elseif strcmp(input, 'HDDA')  || strcmp(input, '7')
    %%
    monomer.name   = 'HDDA';
    monomer.tag    = '7';
    monomer.vmax   = 2;
    monomer.pmax   = 2;
    monomer.nmax   = 2 * monomer.vmax;
    monomer.M      = 226.27;
    monomer.dens   = 1.01;
    monomer.K.R77  = 1;
    monomer.K.R71  = 1;
    monomer.K.R72  = 1;
    monomer.K.R73  = 1;
    monomer.K.R74  = 1;
    monomer.K.R75  = 1;
    monomer.K.R76  = 1;
    monomer.K.init = 6.1197e4;
    monomer.K.prop = 6.1197e4;
    monomer.K.tr   = 4.3955e7;
    monomer.K.td   = 4.3955e7;
else
    disp('This monomer has not yet been added, did you type it right dummy?')
end


if strcmp(type, 'mol')
    monomer.conc     = conc;
elseif strcmp(type, 'g')
    conc_ = conc/monomer.M;
    monomer.conc     = conc_;
else
    disp("That input type is not supported")
end